LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientConnectionManager.cs

```diff
@@ -1,162 +1,399 @@
+                        {
+                        }
+                    });
 
-                    SetupSchedule(managedConnection);
-                }
+                    this.logger.LogDebug($"client connection timeout set to {keepAliveTimeout.TotalSeconds} seconds (sliding expiration)");
+                    return client;
+                });
+            };
 
-                return true;
-            }
+            var key = loRaDevice.DevEUI;
+            if (!this.clientByDevEui.TryAdd(key, loRaDeviceClient))
+                throw new InvalidOperationException($"Connection already registered for device {key}");
+        }
+
+        public void Release(LoRaDevice loRaDevice)
+        {
+            _ = loRaDevice ?? throw new ArgumentNullException(nameof(loRaDevice));
 
-            throw new ManagedConnectionException($"Connection for device {loRaDevice.DevEUI} was not found");
+            if (this.clientByDevEui.TryRemove(loRaDevice.DevEUI, out var removedItem))
+            {
+                removedItem.Dispose();
+            }
         }
 
         /// <summary>
-        /// Sets the schedule for a device client disconnect.
+        /// Tries to trigger scanning of expired items
+        /// For tests only.
         /// </summary>
-        private void SetupSchedule(ManagedConnection managedConnection)
+        public void TryScanExpiredItems()
         {
-            var key = GetScheduleCacheKey(managedConnection.LoRaDevice.DevEUI);
-            // Touching an existing item will update the last access item
-            // Creating will start the expiration count
-            _ = this.cache.GetOrCreate(
-                    key,
-                    (ce) =>
-                    {
-                        ce.SlidingExpiration = TimeSpan.FromSeconds(managedConnection.LoRaDevice.KeepAliveTimeout);
-                        _ = ce.RegisterPostEvictionCallback(OnScheduledDisconnect);
+            _ = this.cache.TryGetValue(string.Empty, out _);
+        }
 
-                        this.logger.LogDebug($"client connection timeout set to {managedConnection.LoRaDevice.KeepAliveTimeout} seconds (sliding expiration)");
+        public void Dispose()
+        {
+            foreach (var client in this.clientByDevEui.Values)
+                client.Dispose();
+        }
 
-                        return managedConnection;
-                    });
+        public IAsyncDisposable BeginDeviceClientConnectionActivity(LoRaDevice loRaDevice)
+        {
+            if (loRaDevice == null) throw new ArgumentNullException(nameof(loRaDevice));
+            return this.clientByDevEui[loRaDevice.DevEUI].BeginDeviceClientConnectionActivity();
         }
 
-        private void OnScheduledDisconnect(object key, object value, EvictionReason reason, object state)
+        private sealed class SynchronizedLoRaDeviceClient : ILoRaDeviceClient, ILoRaDeviceClientSynchronizedOperationEventSource
```

Comment:
no unit tests for this?

Reply:
It's a private class of `LoRaDeviceClientConnectionManager` so it cannot be publicly instantiated. That said, it is unit-tested, albeit indirectly, through the connection manager that returns this client in `GetClient` and the public `ILoRaDeviceClient` members that it implements. See `LoRaDeviceClientConnectionManagerTests`.

Reply:
As discussed. I would have chosen a different take on this and preferred more isolated tests around some of the implementations we have. I see that you did do a lot of tests with the ConnectionManager and I trust you have the cases covered.